src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,575 @@
+ *
+ *    kdfHkdf.deriveKey("AES", kdfParameterSpec);
+ *}
+ *
+ *
+ * @see SecretKey
+ *
+ * @since 23
+ */
+//@PreviewFeature(feature=PreviewFeature.Feature.KEY_DERIVATION)
+public final class KDF {
+    private static final Debug debug = Debug.getInstance("jca",
+                                                         "KDF");
+
+    private static final Debug pdebug = Debug.getInstance("provider",
+                                                          "Provider");
+    private static final boolean skipDebug = Debug.isOn("engine=")
+                                             && !Debug.isOn("kdf");
+
+    // The provider
+    private Provider provider;
+
+    // The provider implementation (delegate)
+    private KDFSpi spi;
+
+    // The name of the MAC algorithm.
+    private final String algorithm;
+
+    // Additional KDF configuration parameters
+    private final AlgorithmParameterSpec algorithmParameterSpec;
+
+    // next service to try in provider selection
+    // null once provider is selected
+    private Service firstService;
+
+    // remaining services to try in provider selection
+    // null once provider is selected
+    private Iterator<Service> serviceIterator;
+
+    private final Object lock;
+
+    /**
+     * Instantiates a KDF object.
+     *
+     * @param keyDerivSpi
+     *     the delegate
+     * @param provider
+     *     the provider
+     * @param algorithm
+     *     the algorithm
+     * @param algParameterSpec
+     *     the algorithm parameters
+     */
+    private KDF(KDFSpi keyDerivSpi, Provider provider, String algorithm,
+                AlgorithmParameterSpec algParameterSpec) {
+        this.spi = keyDerivSpi;
+        this.provider = provider;
+        this.algorithm = algorithm;
+        this.algorithmParameterSpec = algParameterSpec;
+        // the lock is not needed, because the Spi will already be set in
+        // chooseProvider
+        lock = null;
+    }
+
+    private KDF(Service s, Iterator<Service> t, String algorithm,
+                AlgorithmParameterSpec algParameterSpec) {
+        firstService = s;
+        serviceIterator = t;
+        this.algorithm = algorithm;
+        this.algorithmParameterSpec = algParameterSpec;
+        lock = new Object();
+    }
+
+    /**
+     * Returns the algorithm name of this {@code KDF} object.
+     *
+     * <p>This is the same name that was specified in one of the
+     * {@code getInstance} calls that created this {@code KDF} object.
+     *
+     * @return the algorithm name of this {@code KDF} object
+     */
+    public String getAlgorithm() {
+        return this.algorithm;
+    }
+
+    /**
+     * Returns the name of the provider.
+     *
+     * @return the name of the provider
+     */
+    public String getProviderName() {
+        chooseFirstProvider();
+        return provider.getName();
+    }
+
+    /**
+     * Creates an instance of the {@code KDF} object.
```

Comment:
Suggest rewording as "Returns a {@code KDF} object that implements the specified algorithm."